appservice/src/wizard.ts

```diff
@@ -0,0 +1,197 @@
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { AzureAccountWrapper } from './azureAccountWrapper';
+import { SubscriptionModels } from 'azure-arm-resource';
+import { UserCancelledError, WizardFailedError } from './errors';
+
+export type WizardStatus = 'PromptCompleted' | 'Completed' | 'Faulted' | 'Cancelled';
+
+export abstract class WizardBase {
+    private readonly _steps: WizardStep[] = [];
+    private _result: WizardResult;
+
+    protected constructor(readonly output: vscode.OutputChannel) { }
+
+    protected abstract initSteps();
+
+    async run(promptOnly = false): Promise<WizardResult> {
+        this.initSteps();
+
+        // Go through the prompts...
+        for (var i = 0; i < this.steps.length; i++) {
+            const step = this.steps[i];
+
+            try {
+                await this.steps[i].prompt();
+            } catch (err) {
+                this.onError(err, step);
+            }
+        }
+
+        if (promptOnly) {
+            return {
+                status: 'PromptCompleted',
+                step: this.steps[this.steps.length - 1],
+                error: null
+            };
+        }
+
+        return this.execute();
+    }
+
+    async execute(): Promise<WizardResult> {
+        // Execute each step...
+        this.output.show(true);
+        for (var i = 0; i < this.steps.length; i++) {
+            const step = this.steps[i];
+
+            try {
+                this.beforeExecute(step, i);
+                await this.steps[i].execute();
+            } catch (err) {
+                this.onError(err, step);
+            }
+        }
+
+        this._result = {
+            status: 'Completed',
+            step: this.steps[this.steps.length - 1],
+            error: null
+        };
+
+        return this._result;
+    }
+
+    get steps(): WizardStep[] {
+        return this._steps;
+    }
+
+    findStepOfType<T extends WizardStep>(stepTypeConstructor: { new(...args: any[]): T }, isOptional?: boolean): T {
+        return <T>this.findStep(
+            step => step instanceof stepTypeConstructor,
+            isOptional ? null : `The Wizard should have had a ${stepTypeConstructor.name} step`);
+    }
+
+    findStep(predicate: (step: WizardStep) => boolean, errorMessage?: string): WizardStep {
+        const step = this.steps.find(predicate);
+
+        if (!step && errorMessage) {
+            throw new Error(errorMessage);
+        }
+
+        return step;
+    }
+
+    write(text: string) {
+        this.output.append(text);
+    }
+
+    writeline(text: string) {
+        this.output.appendLine(text);
+    }
+
+    private onError(err: Error, step: WizardStep) {
+        if (err instanceof UserCancelledError) {
+            throw err;
+        }
+
+        this.writeline(`Error: ${err.message}`);
```

Comment:
I would let the extensions handle errors instead of printing it here. For example - what if the extension already prints multiline errors to the output window? (Like Nathan just implemented in app service) Then this would show up there twice

@nturinski we need to fix that because it's already an issue in app service haha